x-pack/plugins/fleet/public/applications/fleet/sections/agent_policy/components/agent_policy_advanced_fields/custom_fields/global_data_tags_table.tsx

```diff
@@ -0,0 +1,394 @@
+  const [editIndexList, setEditIndexList] = useState<Set<number>>(new Set([]));
+  // isAdding indicates whether a new row is currently being added to the table.
+  // When true, the table will display "Confirm" and "Cancel" actions for the new row
+  // to allow the user to either confirm the addition or cancel it.
+  const [isAdding, setIsAdding] = useState<boolean>(false);
+  const [newTag, setNewTag] = useState<GlobalDataTag>({ name: '', value: '' });
+  const [newTagErrors, setNewTagErrors] = useState<{ name: string; value: string }>({
+    name: '',
+    value: '',
+  });
+  const [errors, setErrors] = useState<
+    Record<number, { name: string | null; value: string | null }>
+  >({});
+  const [isModalVisible, setIsModalVisible] = useState(false);
+  const [deleteIndex, setDeleteIndex] = useState<number | null>(null);
+
+  // Necessary to remove uncommitted tags after save is cancelled
+  useEffect(() => {
+    setGlobalDataTags(initialTags);
+  }, [initialTags]);
+
+  const handleAddField = () => {
+    setIsAdding(true);
+    setNewTag({ name: '', value: '' });
+  };
+
+  const handleEditChange = (e: React.ChangeEvent<HTMLInputElement>, index: number) => {
+    const updatedTags = globalDataTags.map((tag, i) =>
+      i === index ? { ...tag, [e.target.name]: e.target.value } : tag
+    );
+    setGlobalDataTags(updatedTags);
+    setErrors((prevErrors) => ({
+      ...prevErrors,
+      [index]: { ...prevErrors[index], [e.target.name]: null },
+    }));
+  };
+
+  const handleNewTagChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+    setNewTag({
+      ...newTag,
+      [e.target.name]: e.target.value,
+    });
+    setNewTagErrors({ ...newTagErrors, [e.target.name]: null });
+  };
+
+  const validateTag = (tag: GlobalDataTag, index?: number) => {
+    const trimmedName = tag.name.trim();
+    const trimmedValue = tag.value.toString().trim();
+
+    let nameError = '';
+    let valueError = '';
+
+    if (!trimmedName) {
+      nameError = 'Name cannot be empty';
+    } else if (/\s/.test(trimmedName)) {
+      nameError = 'Name cannot contain spaces';
+    } else if (globalDataTags.some((t, i) => i !== index && t.name === trimmedName)) {
+      nameError = 'Name must be unique';
+    }
+
+    if (!trimmedValue) {
+      valueError = 'Value cannot be empty';
+    }
+
+    return { nameError, valueError, isValid: !nameError && !valueError };
+  };
+
+  const parseValue = (value: string): string | number => {
+    const parsedValue = Number(value);
+    return isNaN(parsedValue) ? value : parsedValue;
+  };
+
+  const confirmNewTagChanges = () => {
+    const { nameError, valueError, isValid } = validateTag(newTag);
+    if (!isValid) {
+      setNewTagErrors({ name: nameError, value: valueError });
+      return;
+    }
+
+    const updatedTags = [
+      ...globalDataTags,
+      { ...newTag, name: newTag.name.trim(), value: parseValue(newTag.value.toString().trim()) },
```

Comment:
Related to what I wrote above, I'm not sure about this line. I think you can avoid casting to a string here, just pass the plain `newTag.value` to `parseValue` and then do a type check.

Reply:
The toString() call is again mainly there for typescript reasons. The value can be either a string or a number, so if I don't have toString() I get the following error 
<img width="804" alt="image" src="https://github.com/elastic/kibana/assets/23665523/1038f280-6dad-4740-8f36-a56d5d3388c7">

Reply:
Typescript gives that error because `trim()` cannot be applied on a `number` type. If you re-write `parseValue` to accept both number and string (see my comment above), then you can then just do:

``` ts
      { ...newTag, name: newTag.name.trim(), value: parseValue(newTag.value) },
```

Does it make sense?